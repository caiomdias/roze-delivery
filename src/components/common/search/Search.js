import React, { useState, useEffect } from 'react'
import { useHistory } from 'react-router-dom'

import GeoCodeAPI from '../../../services/Geocode'

import MapPinIcon from '../../../assets/images/svg/map-pin'
import SearchIcon from '../../../assets/images/svg/search'
import ArrowBack from '../../../assets/images/svg/arrow-back'
import searchImage from '../../../assets/images/png/product-not-found.png'
import addressImage from '../../../assets/images/png/not-found.png'

const SearchInput = ({ type, placeholder }) => {
	const [searchTerm, setSearchTerm] = useState('');
	const [isActive, setActive] = useState(false);
	const [isLoading, setLoading] = useState(false);
	const [results, setResults] = useState([]);
	const [timer, setTimer] = useState(null);
	const [config, setConfig] = useState({ icon: {}, notFound: {} });
	const { icon, notFound } = config;
	const history = useHistory();

	const typeHandle = (searchType) => {
		const types = {
			products: {
				icon: {
					image: <SearchIcon/>,
				},
				notFound: {
					image: searchImage,
					maxWidth: '100px',
					alt: 'Uma lupa sobre a garrafa de cerveja que sumiu',
					message: 'Putz, não conseguimos encontrar o produto'
				}
			},
			address: {
				icon: {
					image: <MapPinIcon/>,
				},
				notFound: {
					image: addressImage,
					maxWidth: '160px',
					alt: 'Uma luneta em cima de um mapa',
					message: 'Ops! Não encontramos seu endereço...'
				}
			}
		};

		setConfig(types[searchType]);
	};

	const backHandle = () => {
		setActive(false);
		setSearchTerm('');
	};

	const fetchAddress = async (value) => {
		const coordenates = await GeoCodeAPI(value);
		const { results } = coordenates;
		const modeledResults = [];

		results.map(
			(
				{
					formatted_address: address,
					geometry: { location }
				}
			) => {
			const obj = {
				address,
				location,
			}
			modeledResults.push(obj);
		})

		setResults(modeledResults);
	};

	const searchHandle = (term) => {
		const { value } = term;
		clearTimeout(timer);

		setTimer(
			setTimeout(() => {
				if (value.length !== 0) {
					if (type === 'address') {
						fetchAddress(value)
					};

					setActive(true);
				} else {
					setActive(false)
				}
			}, 700)
		);

		setSearchTerm(value);
	};

	useEffect(() => {
    	typeHandle(type);
	}, []);

	return (
		<div className="search">
			<div className="search__input">
				<div className="icon">
					{ isActive
						? (<ArrowBack onClick={() => backHandle()} className="back" />)
						: icon.image
					}
				</div>
				<input
					onChange={({ target }) => searchHandle(target)}
					placeholder={placeholder}
					value={searchTerm}
					className="input"
					type="text"
				/>
			</div>

			{
				searchTerm.length !== 0 && (
					<div className={isActive ? 'search__results active' : 'search__results'}>
						{
							results.length !== 0 ? (
								<div className="results">
									{
										results.map(({ address, location }, i) => {
											return (
                        <div key={i} className="search__result-content">
                          <MapPinIcon />
                          <p
                            className="search__result-item"
                            onClick={() => history.push(`/products`, 
                              {
                                lat: location.lat,
                                long: location.lng
                              }
                            )}>{address}
                          </p>
                        </div>
											)
										})
									}
								</div>
							) : (
								<div className="not-found">
									<img
										src={notFound.image}
										alt={notFound.alt}
										style={{maxWidth: notFound.maxWidth}}
									/>
									<p>{notFound.message}
										<strong>
										{
											type !== 'address' && searchTerm
										}
										</strong>
									</p>
								</div>
							)
						}
					</div>
				)
			}

		</div>
	)
};

SearchInput.defaultProps = {
	type: 'products',
	placeholder: '',
};

export default SearchInput;
