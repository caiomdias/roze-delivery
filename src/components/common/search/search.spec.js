import React from 'react';
import renderer from 'react-test-renderer';
import Search from './Search';

const config = {
	type: 'products',
	placeholder: 'test',
}

describe("Search Component", () => {
	it('render correctly Search component', () => {  
		const searchComponent = renderer.create(<Search type={config.type} placeholder={config.placeholder}/>);
		expect(searchComponent).toMatchSnapshot();
	});
});