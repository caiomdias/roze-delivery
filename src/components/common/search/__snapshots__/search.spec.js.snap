// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component render correctly Search component 1`] = `
<div
  className="search"
>
  <div
    className="search__input"
  >
    <div
      className="icon"
    />
    <input
      className="input"
      onChange={[Function]}
      placeholder="test"
      type="text"
      value=""
    />
  </div>
</div>
`;
